modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/certificatevalidation/CertificateVerificationManager.java

```diff
@@ -0,0 +1,263 @@
+
+                        // Store the valid issuer cert in cache for future use
+                        certCache.setCacheValue(peerCert.getSerialNumber().toString(), issuerCert);
+                        if (log.isDebugEnabled()) {
+                            log.debug("Issuer certificate with serial number: " + issuerCert.getSerialNumber()
+                                    .toString() + " has been cached against the serial number:  " + peerCert
+                                    .getSerialNumber().toString() + " of the peer certificate.");
+                        }
+                        break;
+                    } catch (SignatureException | CertificateException | NoSuchAlgorithmException |
+                             InvalidKeyException | NoSuchProviderException e) {
+                        // Unable to verify the signature. Check with the next certificate.
+                    }
+                }
+            } else {
+                X509Certificate cachedIssuerCert = certCache.getCacheValue(peerCert.getSerialNumber().toString());
+                try {
+                    peerCert.verify(cachedIssuerCert.getPublicKey());
+                } catch (SignatureException | CertificateException | NoSuchAlgorithmException |
+                         InvalidKeyException |
+                         NoSuchProviderException e) {
+                    // Unable to verify the signature.
+                    throw new CertificateVerificationException("Unable to verify the signature of the certificate.");
+                }
+            }
+        }
+
+        long start = System.currentTimeMillis();
+
+        OCSPCache ocspCache = OCSPCache.getCache();
+        ocspCache.init(cacheSize, cacheDelayMins);
+        CRLCache crlCache = CRLCache.getCache();
+        crlCache.init(cacheSize, cacheDelayMins);
+
+        RevocationVerifier[] verifiers = {new OCSPVerifier(ocspCache), new CRLVerifier(crlCache)};
+
+        for (RevocationVerifier verifier : verifiers) {
+            try {
+                if (isFullCertChainValidationEnabled) {
+                    log.debug("Doing full certificate chain validation");
+                    CertificatePathValidator pathValidator = new CertificatePathValidator(convertedCertificates,
+                            verifier);
+                    pathValidator.validatePath();
+                    log.info("Path verification Successful. Took " + (System.currentTimeMillis() - start) + " ms.");
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Validating client certificate with the issuer certificate retrieved from" +
+                                "the trust store");
+                    }
+                    verifier.checkRevocationStatus(peerCert, issuerCert);
+                }
+                return;
+            } catch (Exception e) {
```

Comment:
check this

Reply:
This is not handled here since the next verifier should be executed if one verifier is failed. If both verifiers fail, a ```CertificateVerificationException``` is thrown in line number 209.